[jit] Add filter function to subgraph rewriter runGraph

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26224 [quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer
* **#26223 [jit] Add filter function to subgraph rewriter runGraph**

Summary:
add filter function to runGraph, if the function returns false for given `Match`,
the we'll skip the rewrite.

Test Plan:
will test in later PR that adds extra filtering on Constant nodes

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17462529](https://our.internmc.facebook.com/intern/diff/D17462529)

jerryzh168 committed:
[jit] Add filter function to subgraph rewriter runGraph

Summary:
add filter function to runGraph, if the function returns false for given `Match`,
the we'll skip the rewrite.

Test Plan:
will test in later PR that adds extra filtering on Constant nodes

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[jit] Add filter function to subgraph rewriter runGraph"


Summary:
add filter function to runGraph, if the function returns false for given `Match`,
the we'll skip the rewrite.

Test Plan:
will test in later PR that adds extra filtering on Constant nodes

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

ZolotukhinM commented on torch/csrc/jit/passes/subgraph_rewrite.h:
```diff
@@ -47,7 +48,16 @@ class TORCH_API SubgraphRewriter {
   // Run pattern-based subgraph rewrite pass on the graph (used in testing).
-  void runOnGraph(std::shared_ptr<Graph>& graph);
+  // filter is a function that does extra filtering on the match, if it returns
+  // true for a given Match, we'll skip the match
```
The comment contradicts with the code: we skip if the filter returns false.

Also, please add a detailed description of the arguments of this function - what is the meaning of its arguments and how in a usual case users of this API could obtain them.

ZolotukhinM reviewed (changes requested):
The code and API look good, but changes to such components **must** have tests (note that the tests should be generic and not quantization-specific).

jerryzh168 committed:
Update on "[jit] Add filter function to subgraph rewriter runGraph"


Summary:
add filter function to runGraph, if the function returns false for given `Match`,
the we'll skip the rewrite.

Test Plan:
will test in later PR that adds extra filtering on Constant nodes

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add filter function to subgraph rewriter runGraph"


Summary:
add filter function to runGraph, if the function returns false for given `Match`,
the we'll skip the rewrite.

Test Plan:
will test in later PR that adds extra filtering on Constant nodes

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/jit/test_subgraph_rewriter.cpp:
```diff
@@ -0,0 +1,110 @@
+      &pattern_graph,
+      vmap);
+
+  auto matches = findPatternMatches(pattern_graph, *graph);
```
This is unused.

ZolotukhinM commented on test/cpp/jit/test_subgraph_rewriter.cpp:
```diff
@@ -0,0 +1,110 @@
+      &pattern_graph,
+      vmap);
+
+  auto matches = findPatternMatches(pattern_graph, *graph);
```
Ditto.

ZolotukhinM reviewed (approved):
Looks good, thanks!

jerryzh168 committed:
Update on "[jit] Add filter function to subgraph rewriter runGraph"


Summary:
add filter function to runGraph, if the function returns false for given `Match`,
the we'll skip the rewrite.

Test Plan:
will test in later PR that adds extra filtering on Constant nodes

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
